include/seastar/core/metrics.hh

```diff
@@ -357,7 +349,23 @@ public:
                 ((_type == data_type::COUNTER || _type == data_type::REAL_COUNTER) && d() == 0);
     }
 private:
-    static void ulong_conversion_error(double d);
+    static void ulong_conversion_error(double d, const char* ctype);
+
+    template <typename T>
+    T dtoi(const char* ctype) const noexcept {
+        auto d = std::get<double>(u);
+        if (d <= double(std::numeric_limits<T>::min())) [[unlikely]] {
+            ulong_conversion_error(d, ctype);
+            return std::numeric_limits<T>::min();
+        }
+        // Limit maximum value to the signed maximum
+        // since lround converts to signed long
+        if (d >= double(std::numeric_limits<int64_t>::max())) [[unlikely]] {
```

Comment:
i observe three issues here.

1. if the value of `d` is equal to `std::numeric_limits<int64_t>::max()`, it should be accepted.
2. to convert `std::numeric_limits<int64_t>::max()` to double using `double(std::numeric_limits<int64_t>::max())` changes its value.  in other words, `double()` rounds a large integer to the nearest float point number which can be represented by `double`. in this case, before rounding, the value is changed from $1^{63}-1$ to $1^{64}$. so we are more permissive than we should by enlarging the upper bound by 1. this is probably not a big deal, but IMHO, is wrong. i have a fix demonstrated at https://godbolt.org/z/oe547afjj . FWIW, which is included as part of https://github.com/scylladb/scylladb/pull/12874 
3. IMHO, to truncate the number to the maximum value returned by `lround()` even if `T` here could represent a greater value, is wrong. we can cast double to `T` and `cmp()` them. but i admit we probably need to document `cmp()` in a easier-to-digest way.

Reply:
interesting direction.
but it seems to be unsymmetrical, see https://godbolt.org/z/b7bYW1o6x

Reply:
the unsymmetrical behavior is expected and correct, i think. as double is able to represent `numeric_limits<int64_t>::min()`.

Reply:
@tchaikov as I noted in the other thread, there is no way to "fix" this issue. If the number is already held as a  "double" (here, in `std::get<double>(u)`), it already lost the ability to tell `2^63-1` from `2^63`. There is absolutely nothing you can "fix" in the code to tell these cases apart, because they are already mapped to the same double.

We need to concede that if our metrics library keeps "gauges" as doubles, when the magnitude of the counter grows, you can no longer make small increments to it. For example, if the counter reaches `2^53`, adding one to it will not change it any more! (this is known as DBL_EPSILON, see also https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon) The counter can only continue to increase if you add larger increments to it. At `2^63-1` differences of 1 (or even up to 1024.0) are lost, so it is meaningless to ask that `2^63-1` be treated differently from `2^63` because they are the same number.

I agree that we should avoid cases that result in floating-point exception or ASAN warnings, but other than that, we can't really "fix" this case to be perfect in any way. If counters grow beyond `2^53` (1/DBL_EPSILON), converting them to in an integer will be inaccurate in any case, and dubious. This brings us back to @amnonh's comment - of why we reached such high counters in the first place, or adding to the question myself: If they are not "counters" and genuinely double-precision sums of something of potentially large manitudes, why are we converting them to limited-magnitude integers?

Reply:
We might as well change the variant in metric_value to also hold int64_t and uint64_t. This will improve precision for counters that hit 2^53. That's 8PB, so even at 1GB/s will need 8M seconds to reach.

@amnonh ?

Reply:
@amnonh please comment on @avikivity's proposal above :point_up: